### PR TITLE
Support mutable state by stashing commands while storing snapshot, #25740

### PR DESCRIPTION
**Pull request by patriknw**
Stashing while storing snapshot is easy to reason about. We could consider optimization opt-in for immutable state later if it turns out to be important to support completely async snapshotting. 

Refs #25740

### CODE REVIEW

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (patriknw) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]: created separate issue for missing api https://github.com/akka/akka/issues/26273

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (johanandren) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]: ".."

File: akka-docs/src/main/paradox/typed/persistence.md
Comment (johanandren) [inline akka-docs/src/main/paradox/typed/persistence.md]: Should this sentence have stronger wording about not doing side effects?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala]: ticket?

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala
Comment (johanandren) [inline akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala]: Am I missing how this test somehow verifies ordering, shouldn't Increment cmd and evt handler tell the probe so that we can verify that 4, 5 happened after the snapshot-success?

**Review by johanandren — COMMENTED** [review]
Looks good, some minor comments

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala
Comment (patriknw) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala]: it did by checking for concurrent modifications but that was very vague and implicit. I have improved the test in 5c31502

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala]: done, was existing ticket

Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala]: added number to FIXME

File: akka-docs/src/main/paradox/typed/persistence.md
Comment (patriknw) [reply akka-docs/src/main/paradox/typed/persistence.md]: rephrase

File: akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/BasicPersistentBehaviorCompileOnly.scala
Comment (helena) [inline akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/BasicPersistentBehaviorCompileOnly.scala]: very helpful, and particularly so per data type

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]: This is one of my concerns, that it is only by number of events versus snapshot/batch size or time interval options. Would be nice to have a few options available, rather than number of events. Batch size and snapshot size allows are more easily calculable sizing in parallel with provisioning, particularly by data set.

Comment (helena) [reply akka-docs/src/main/paradox/typed/persistence-snapshot.md]: Which I am noting here but it's not relevant to this change of course. I have it noted elsewhere.

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]:
```suggestion
When a snapshot is triggered, incoming commands are stashed until the snapshot has been saved. This means that
```

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]: I'd make this not a run-on sentence.

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]: When we say 'previously saved' do we mean if the `keep` is say set to 2, the system snapshots 2 batches plus the fresh one. And 'using' this most recent of that?

Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]:
```suggestion
During recovery, the persistent actor is using the previously saved snapshot to initialize the state. And events
```
Or the sentence needs some other clarification/breakup if this isn't what you mean.

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]: What about `deleteMessages` from `EventSourced` with snapshotting? Should we clarify this last sentence?

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]:
```suggestion
Since it is acceptable for some applications to not use any snapshotting, it is perfectly valid not to configure a snapshot store.
```

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence-snapshot.md]:
```suggestion
the `onSnapshot` callback. Snapshot failures are, by default, logged but do not cause the actor to stop or
```

File: akka-docs/src/main/paradox/typed/persistence.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence.md]: I'd clarify this. If it is immutable 
```suggestion
The state is immutable, the event handler returns a new instance of the state. If the state is
```
And clarify when or why state is not immutable.

File: akka-docs/src/main/paradox/typed/persistence.md
Comment (helena) [inline akka-docs/src/main/paradox/typed/persistence.md]:
```suggestion
The event handler should only update the state and never perform side effects, as those would also be
```

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/JournalInteractions.scala
Comment (helena) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/JournalInteractions.scala]:
```suggestion
      throw new IllegalStateException("A snapshot must not be a null state.")
```

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/JournalInteractions.scala
Comment (helena) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/JournalInteractions.scala]: We declare return Unit but in fact it could throw. Should this be documented?

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala
Comment (helena) [inline akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala]: Is there another check that can be done vs use this?

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (helena) [reply akka-docs/src/main/paradox/typed/persistence-snapshot.md]:
```suggestion
TODO #26273 include corresponding example in Java
```

**Review by helena — CHANGES_REQUESTED** [review]
Great start, perhaps a few minor clarifications

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (patriknw) [reply akka-docs/src/main/paradox/typed/persistence-snapshot.md]: changed "previously saved" to "latest saved", "fresh" also refers to "latest" so changed that as well

Comment (patriknw) [reply akka-docs/src/main/paradox/typed/persistence-snapshot.md]: changed to "initialize the state. Thereafter the events after the snapshot are replayed"

File: akka-docs/src/main/paradox/typed/persistence-snapshot.md
Comment (patriknw) [reply akka-docs/src/main/paradox/typed/persistence-snapshot.md]: "events have been deleted" corresponds to `deleteMesages`. API for that hasn't been added in Typed yet, issue https://github.com/akka/akka/issues/24698

Note that in untyped persistence, the events are often called messages, which is probably confusing. In Typed we should consistently call them events.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/JournalInteractions.scala
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/JournalInteractions.scala]: `IllegalStateException` here would be a bug in Akka, we should not attempt to save null snapshots

File: akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala
Comment (patriknw) [reply akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/SnapshotMutableStateSpec.scala]: then it would have to pass in a CountDownLatch in the snapshot, but that seems complicated here since the purpose of the test is to exercise race conditions

**Review by patriknw — COMMENTED** [review]
I have incorporated feedback from @helena 
It was all about (good) doc improvements so I don't think we have to wait for a re-review before merging this.

**Pull request merged**
